### ISSUE TITLE
_Process double type delta time introduces casting inconveniences

### ISSUE THREAD

**Issue by dpalais** [godotengine/godot#65313]
### Godot version

4.0.dev (b6d102c)

### System information

Fedora Linux 36

### Issue description

The change of node processing delta time to double, recently reflected in C# scripts, has surfaced some compilation errors that are fixable with new explicit downcasts to float:

AnimationTree.Advance(delta) : `delta` is float
Mathf methods don't accept parameters that are the result of arithmetical operations with double `delta` without downcasting

While this can be fixed with a few extra keywords I only wanted to have you consider this new situation.

### Steps to reproduce

```csharp
using Godot;

public partial class TestCase : Node3D
{
  public TestCase()
  {
    AddChild(animationTree);
  }

  public override void _Process(double delta)
  {
    var myVar = delta * 100;
    RotateObjectLocal(Vector3.Up, myVar); // error
    var myVar2 = Mathf.Cos(myVar); // error
    animationTree.Advance(delta); // error
  }

  private AnimationTree animationTree = new AnimationTree();
}
```

**Comment by raulsntos** [comment]
Using `float` as the type of the delta parameter is wrong because this type is declared as `double` in core.

https://github.com/godotengine/godot/blob/b6d102c7c2d5e9f39f368bf1ebc91a48bf762c42/scene/main/node.h#L237

**Comment by raulsntos** [comment]
Using `float` as the type of the delta parameter is wrong because this type is declared as `double` in core.

https://github.com/godotengine/godot/blob/b6d102c7c2d5e9f39f368bf1ebc91a48bf762c42/scene/main/node.h#L237

**Comment by dpalais** [comment]
I just noticed that AnimationTree properly uses double time internaly so in this case it's only a matter of fixing the public facing API to accept double too.

That is or was the most important problem. The others might be inconveniences but still interesting to discuss.

**Comment by akien-mga** [comment]
Why does `AnimationTree.Advance` take a float in C# when it's double internally / in GDScript?

The only single precision floats in the Godot API should be in vector/matrix types.
For `Mathf`, the `f` means float, I assume C# has a `Math` or `Mathd` for double precision.

**Comment by raulsntos** [comment]
> Why does `AnimationTree.Advance` take a float in C# when it's double internally / in GDScript?

`AnimationTree::advance` is declared with a parameter of type `real_t` in C++:

https://github.com/godotengine/godot/blob/e7a0a97c0bc5d86644e62d537503e3b05313a01c/scene/animation/animation_tree.h#L354

This means it will be generated as `float` in C# unless compiled with `--float=64`. If it's supposed to always be a `double` changing it in core will also change it in the C# bindings.

> For `Mathf`, the `f` means float, I assume C# has a `Math` or `Mathd` for double precision.

In C#, the BCL has a static [`System.Math`](https://docs.microsoft.com/en-us/dotnet/api/system.math?view=net-6.0) class that uses double and a static [`System.MathF`](https://docs.microsoft.com/en-us/dotnet/api/system.mathf?view=net-6.0) class that uses float.

In Godot we have our own static `Godot.Mathf` class that uses `real_t` so it will use float unless compiled with `--float=64`.

**Comment by dpalais** [comment]
So far I have found other classes that also have headers with `float` time:
**AnimationNodeOneShot** `set_fadein_time` `set_fadeout_time`
**AnimationNodeTransition** `set_xfade_time`
**AnimationPlayer** `seek_delta` `advance`
**AnimationTree** `advance`
**AudioStream\***  `seek`
**SceneTree** `process_timers` `process_tweens`
**Tween** `step` `custom_step`
**VideoStreamPlayback** `update`
**VideoStreamPlaybackTheora** `update` and class members

**Comment by raulsntos** [comment]
Those might need to be updated.
cc @aaronfranke

**Comment by aaronfranke** [comment]
In general, we should use `double` for methods that deal with time, not `real_t`. The precision of `float` is not enough to do calculations for more than a few hours, and errors will build up over time. Even if these methods don't need it, it's good to use `double` for time for consistency with the other places that do need `double` for time. So I think it would be good to change all of the methods that @dpalais found to use `double`.

**Comment by dpalais** [comment]
There's a pull request that addresses this at #66160

**Comment by akien-mga** [comment]
Fixed by #66160.

**Comment by GeekyMonkey** [comment]
The problem with this is that the time delta is often used in computing position translations.  If we try to do all of this stuff using doubles, then we have the cast problem when trying to create a Vector 2

```
    double newX = Position.X + input * MoveSpeed * (float)delta;
    newX = Mathf.Clamp(newX, XMin, XMax);
    Position = new Vector2(newX, Position.Y);
```
        
Error: Argument 1 cannot convert from double to float
        
It would be nice if we could get the time delta as a float, since 99% of what we do only needs float precision.
